### PR TITLE
Implement the Suggestions Database

### LINKED ISSUE #829: Implement the Suggestions Database
### Summary
The Suggestions database holds entries for fulfilling the auto-completion requests and contains additional indexes used to answer the different search criteria.

### Value
The database contains entries in a suitable format that allows us to quickly respond to autocomplete requests.

### Specification

- [ ] Implement the code and data structures required to back the suggestions database.
- [ ] Implement a traversal over the IR that gathers the requisite information and fills the database using it.

### Acceptance Criteria & Test Cases

- The database is implemented.
- The module contains methods to fill the database from a given `IR`.

### PR DESCRIPTION
**Pull request by 4e6**
### Pull Request Description

close #829

Changelog:

- add: new `searcher` sbt project
- add: SQLite backend for the suggestions database
- add: `SuggestionBuilder` that extracts the suggestions from a given `IR`

### Checklist
Please include the following checklist in your PR:

- [x] The documentation has been updated if necessary.
- [x] All code conforms to the [Scala](https://github.com/luna/enso/blob/main/docs/style-guide/scala.md), [Java](https://github.com/luna/enso/blob/main/docs/style-guide/java.md), and [Rust](https://github.com/luna/enso/blob/main/docs/style-guide/rust.md) style guides.
- [x] All code has been tested where possible.

### CODE REVIEW

File: lib/searcher/src/test/scala/org/enso/searcher/sql/SuggestionsDBIOTest.scala [hunk 1]
```diff
@@ -0,0 +1,81 @@
...
+        for {
+          _   <- db.run(dbio.insert(stub.local))
+          res <- db.run(dbio.findBy(stub.local.returnType))
+        } yield res
+
+      Await.result(action, Timeout) shouldEqual Seq(stub.local)
+    }
+  }
+
+  object stub {
```
Comment (lolczak) [inline lib/searcher/src/test/scala/org/enso/searcher/sql/SuggestionsDBIOTest.scala:53]: It not a stub. A stub is used for providing the tested code with "indirect input".

File: lib/searcher/src/test/scala/org/enso/searcher/sql/SuggestionsDBIOTest.scala [hunk 2]
```diff
@@ -0,0 +1,81 @@
...
+    "findBy returnType" in {
+      val action =
+        for {
+          _   <- db.run(dbio.insert(stub.local))
+          res <- db.run(dbio.findBy(stub.local.returnType))
+        } yield res
+
+      Await.result(action, Timeout) shouldEqual Seq(stub.local)
+    }
+  }
```
Comment (lolczak) [inline lib/searcher/src/test/scala/org/enso/searcher/sql/SuggestionsDBIOTest.scala:51]: Please specify behaviors better than single-word sentences and test it more thoroughly for all possible cases.

File: lib/searcher/src/main/scala/org/enso/searcher/sql/SuggestionsDBIO.scala [hunk 3]
```diff
@@ -0,0 +1,186 @@
...
+
+  /** Find suggestions by the return type.
+    *
+    * @param returnType the return type of a suggestion
+    * @return the list of suggestions
+    */
+  def findBy(returnType: String): DBIO[Seq[Suggestion]] = {
+    val q = for {
+      (a, s) <- joined
+      if s.returnType === returnType
```
Comment (lolczak) [inline lib/searcher/src/main/scala/org/enso/searcher/sql/SuggestionsDBIO.scala:30]: Please don't use single-letter variable names. Fix here and other places.

File: lib/searcher/src/main/scala/org/enso/searcher/sql/SuggestionsDBIO.scala [hunk 4]
```diff
@@ -0,0 +1,186 @@
+package org.enso.searcher.sql
+
+import org.enso.searcher.Suggestion
+import slick.jdbc.SQLiteProfile.api._
+
+import scala.concurrent.ExecutionContext
+
+/** The object for accessing the suggestions database. */
+final class SuggestionsDBIO(implicit ec: ExecutionContext) {
```
Comment (lolczak) [inline lib/searcher/src/main/scala/org/enso/searcher/sql/SuggestionsDBIO.scala:9]: The object for accessing a database is called a repository. I suggest renaming it.

Comment (lolczak) [reply lib/searcher/src/main/scala/org/enso/searcher/sql/SuggestionsDBIO.scala:9]: It is a good practice to extract interface for a repository to loosely couple clients with a concrete implementation.

File: lib/searcher/src/test/scala/org/enso/searcher/sql/SuggestionsDBIOTest.scala (hunk 2 above)
Comment (4e6) [reply lib/searcher/src/test/scala/org/enso/searcher/sql/SuggestionsDBIOTest.scala:51]: I extended the scenarios a bit. There will be more methods (and tests) when I start integrating the database with the runtime.

File: build.sbt [hunk 5]
```diff
@@ -570,6 +570,18 @@ lazy val `core-definition` = (project in file("lib/core-definition"))
   .dependsOn(graph)
   .dependsOn(syntax.jvm)
 
+lazy val searcher = project
+  .in(file("lib/searcher"))
+  .configs(Test)
+  .settings(
+    Compile / console / scalacOptions -= "-Xfatal-warnings",
```
Comment (iamrecursion) [inline build.sbt:577]: Why are you turning off fatal warnings?

File: engine/runtime/src/main/scala/org/enso/interpreter/instrument/SuggestionBuilder.scala [hunk 6]
```diff
@@ -0,0 +1,162 @@
+package org.enso.interpreter.instrument
```
Comment (iamrecursion) [inline engine/runtime/src/main/scala/org/enso/interpreter/instrument/SuggestionBuilder.scala:1]: Why is this under instrument? It isn't an instrument.

File: engine/runtime/src/test/scala/org/enso/compiler/test/context/SuggestionBuilderTest.scala [hunk 7]
```diff
@@ -0,0 +1,310 @@
...
+import org.enso.interpreter.instrument.SuggestionBuilder
+import org.enso.searcher.Suggestion
+
+class SuggestionBuilderTest extends CompilerTest {
+
+  implicit val passManager: PassManager = new Passes().passManager
+
+  "SuggestionBuilder" should {
+
+    "build method without arguments" in {
```
Comment (iamrecursion) [inline engine/runtime/src/test/scala/org/enso/compiler/test/context/SuggestionBuilderTest.scala:17]: Strictly speaking, at the top level this is not a method without arguments. You even check that the `this` argument exists below!

File: engine/runtime/src/test/scala/org/enso/compiler/test/context/SuggestionBuilderTest.scala [hunk 8]
```diff
@@ -0,0 +1,310 @@
...
+      )
+    }
+  }
+
+  private def build(ir: IR.Module): Vector[Suggestion] =
+    new SuggestionBuilder().build(ir)
+
+  private def freshModuleContext: ModuleContext =
+    ModuleContext(freshNameSupply = Some(new FreshNameSupply))
+}
```
Comment (iamrecursion) [inline engine/runtime/src/test/scala/org/enso/compiler/test/context/SuggestionBuilderTest.scala:310]: What's missing here are the following:

- Getting type information from associated type signatures.

File: build.sbt (hunk 5 above)
Comment (4e6) [reply build.sbt:577]: They were removed only for the `console` task

**Comment by 4e6** [comment]
Changelog:

- add: support for type signatures
- fix: review comments

**Pull request merged**